### DISCUSSION TITLE
Open-source inquires

### DISCUSSION THREAD

**Discussion by abuibaid** [coin-or/Ipopt#680] [General]
Hi everyone,
We have developed software which is dependent on CasADI and IPOPT. We need to publish the code here in github with licesne. I know CasADI's license has a copyleft, as well as the ipopt which has weak copyleft. Actually there is no modification occured on whether CasADI or IPOPT, while the aim is to redistribute the software which is dependent on these tools. The questions are : 1) how much freedom do I have without violating the copyleft? 2) are there any compatible licenses with CasADI or I can choose whatever I want?

I hope you guys explain this to me with simple terms :D :D :D

Regards,

**Comment by svigerske** [comment]
If you just want to publish a copy of Ipopt, then that is no problem. Just keep the `LICENSE` file, so people know that the Ipopt part is under Eclipse Public License.

You are also free to do changes. In that case it would be good to specify somewhere (a readme or so), where you did some changes.

If you need to publish Ipopt under a different license, then that seems difficult. From what I read in https://github.com/coin-or/Ipopt/blob/stable/3.14/LICENSE#L136-L146, the Ipopt license would have to specify additional licenses explicitly, but it doesn't do so.

I cannot say anything about CasADI.

**Reply by tkralphs** [reply]
CasADI is under ~GPL~ LGPL v3 and Ipopt is under the CPL. You almost surely cannot change the license on either of these and the licenses may be incompatible with respect to distributing binaries (the LGPL allows for the redistribution of some kinds of combinations in binary form). Redistributing sources should be allowed by both licenses, ~although there is some subtlety around non-GPL'd code using the API of GPL'd code. I guess that is not the case here.~ Bottom line: you ~almost certainly~ probably cannot distribute a combination in binary form, but source form is probably OK.

**Reply by traversaro** [reply]
Are you sure CasADi is GPL v3 ? From the repo it states it is licensed under the LGPL v3 license: https://github.com/casadi/casadi/blob/main/LICENSE.txt .

**Reply by tkralphs** [reply]
>Are you sure CasADi is GPL v3 ? From the repo it states it is licensed under the LGPL v3 license

Yes, you are correct, I was too hasty. Whether this matters to the answer above depends on exactly how CasADI and Ipopt are combined in the final product. Obviously, consulting a lawyer and not taking advice from amateurs would be a good idea :).

**Reply by abuibaid** [reply]
Consulting a lawyer would be really the eventual solution :D.  But the thing that, CasADI is using IPOPT as their solver and if you want to publish software that is using CasADI including IPOPT, there is conflict between CasADI licenses and IPOPT licenses in first place, as well as another conflict if you are using another complete different license for you own software!! WOAH, this thing needs for real a lawyer!

Anyway thanks guys for explaining this tought matter <3
